### PR TITLE
Preserves requires_grad state in distribute_module and tensor_parallel

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #171523 

### These changes fix a bug where requires_grad state was lost when converting parameters to distributed tensors. 

**Problem**: 
When distribute_module or tensor parallel functions created new nn.Parameter objects from distributed tensors, frozen parameters (with requires_grad=False) would incorrectly become trainable. 

**Fix**: 
Pass `requires_grad=param.requires_grad` explicitly when creating nn.Parameter objects in:

- `distribute_module` in _api.py
- `ColwiseParallel._partition_linear_fn` and `_partition_embedding_fn` in style.py
- `RowwiseParallel._partition_linear_fn` and `_partition_embedding_fn` in style.py

**Tests:** 
Added three simple tests verifying frozen parameters stay frozen and trainable parameters stay trainable after parallelization.

### CODE REVIEW

File: torch/distributed/tensor/_api.py [hunk 1]
```diff
@@ -1000,7 +1000,10 @@ def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
             if param is not None and not isinstance(param, DTensor):
                 m.register_parameter(
                     key,
-                    nn.Parameter(distribute_tensor(param.data, mesh, full_replicate)),
+                    nn.Parameter(
+                        distribute_tensor(param.data, mesh, full_replicate),
+                        requires_grad=param.requires_grad,
+                    ),
```
Comment (tianyu-l) [inline torch/distributed/tensor/_api.py:1061]: It seems `distribute_tensor` itself will respect plain tensor's `requires_grad`.
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_api.py#L877

The "unexpected" behavior is coming from nn.Parameter by default sets `requires_grad` to True.
https://github.com/pytorch/pytorch/blob/v2.9.1/torch/nn/parameter.py#L51

@albanD would like to hear your take on this issue
- why nn.Parameter has this behavior in the first place?
- should we explicitly alter nn.Parameter behavior in DTensor APIs?

Comment (arkadip-maitra) [reply torch/distributed/tensor/_api.py:1061]: True. `nn.Buffer` already preserves the input's requires_grad, but `nn.Parameter` for custom tensors doesn't, applies the default True.
Aligning `nn.Parameter` custom tensor path with `nn.Buffer` would be the actual fix but that needs to be decided

Comment (arkadip-maitra) [reply torch/distributed/tensor/_api.py:1061]: Hi @tianyu-l should I make any updates on this? changing `nn.Parameter` to not default to True would be easy change but that would probably break a lot of existing code. any idea on what should be done?

Comment (tianyu-l) [reply torch/distributed/tensor/_api.py:1061]: we need @albanD , let me ask him directly

File: torch/distributed/tensor/_api.py
Comment (albanD) [inline torch/distributed/tensor/_api.py:1065]: Unrelated: note that there is a nn.Buffer that exists https://docs.pytorch.org/docs/stable/generated/torch.nn.parameter.Buffer.html#buffer that has a similar behavior to Parameter. You might want to add support here.

File: torch/distributed/tensor/_api.py (hunk 1 above)
Comment (albanD) [reply torch/distributed/tensor/_api.py:1061]: > why nn.Parameter has this behavior in the first place?

This is because for all Tensors across the board, by default, they do not require gradients (creating input, buffers, stats) and only the model Parameters require gradients by default.
The logic here is really that if you create a fresh new Parameter, then you are going to set it as the Parameter on your Module and thus it should require gradient (expect very specific cases).

> should we explicitly alter nn.Parameter behavior in DTensor APIs?

Not sure what you mean by this.
But the change here sounds good to me: you're not creating a Parameter from scratch, you're replicating and existing one. So you are copying over the properties of the original one instead of taking the default values.

File: test/distributed/tensor/test_api.py [hunk 2]
```diff
@@ -287,6 +287,26 @@ def replicate_input_fn(mod, inputs, device_mesh):
...
+
+            def forward(self, x):
+                return x + self.frozen + self.trainable
+
+        model = ModelWithFrozenParam().to(self.device_type)
+
+        distributed_model = distribute_module(model, device_mesh)
+
+        self.assertFalse(distributed_model.frozen.requires_grad)
+        self.assertTrue(distributed_model.trainable.requires_grad)
```
Comment (albanD) [inline test/distributed/tensor/test_api.py:308]: nit: another test I would add here would be to do a forward with a dummy input, then .sum().backward() and check that the .grad field is only populated for trainable.
That way we're sure that even the intermediaries that are being re-created don't end up computing gradients for frozen.

**Review by albanD — APPROVED** [review]
SGTM!

File: test/distributed/tensor/test_api.py (hunk 2 above)
Comment (arkadip-maitra) [reply test/distributed/tensor/test_api.py:308]: updated this test. thanks for the suggestion

File: torch/distributed/tensor/_api.py
Comment (arkadip-maitra) [reply torch/distributed/tensor/_api.py:1065]: will make this change in another pr. thanks

**Review by wconstab — APPROVED** [review]

**Review by tianyu-l — APPROVED** [review]
